### PR TITLE
Feature/react ts client preview

### LINKED ISSUE #140: Preview app does not support datasets where min == max
Activating the layer does not work. I get

```
Error: noUiSlider (12.0.0): 'range' 'min' and 'max' cannot be equal.
```

in the JS console. Seems like we will have to catch this case...

### LINKED ISSUE #195: Rewrite TC preview app
For a better UI, maintainability, and consistency, the TC preview app can be re-written in React and TS.

The interface can be similar to the apps we keep building such as [satlas.dk](satlas.dk) or [urban.satlas.dk](urban.satlas.dk).

### PR DESCRIPTION
**Pull request by bertearazvan**
This PR aims to rewrite the Terracotta preview app using React and Typescript based on the create-react-app boilerplate.
- it adds the webpack configuration for local development and production build.
- it adds the flask script to run the app
- it recreates the app using the same kind of UI as satlas.dk

Close #195 
Close #140

### CODE REVIEW

**Comment by bertearazvan** [comment]
d6aeeff33e43970ba3e26227f36d7fa2

**Comment by dionhaefner** [comment]
Are you planning to address some of our long-standing issues? #91 #135 #140 #182

**Comment by bertearazvan** [comment]
Probably #140 is fixed(currently you can only slide, and not use an input for those min/max values - unsure if this is an issue. The precision of the slider for singleband values is `0.01`, not sure if it should be more precise than this) and I'll look into #182, the other two seem a bit complex for this PR. 

There are discussions going around #91 about having a client library that wraps TC with a Map provider - probably through deck.gl. Stay tuned

As about #135, hmm interesting feature, sparks my interest so I might look into it, but as a separate PR.

**Comment by dionhaefner** [comment]
OK. I personally prefer entering values to sliding, but your call if you want to support that.

#140 was not about this, though. The bug arose when a raster had a constant value, e.g. all 0. Then, the slider would break. So you should probably test this case to make sure it's fixed.

File: terracotta/client/app/src/App.tsx [hunk 1]
```diff
@@ -82,7 +82,7 @@ const App: FC<Props> = ({ hostnameProp }) => {
 
   const initializeApp = (hostname: string | undefined) => {
     // sanitize hostname
-    // hostname = 'https://4opg6b5hc3.execute-api.eu-central-1.amazonaws.com/development'
+    hostname = 'https://4opg6b5hc3.execute-api.eu-central-1.amazonaws.com/development'
```
Comment (j08lue) [inline terracotta/client/app/src/App.tsx:85]: Is this only used during development? We should probably not include this in the code - or use an "official" demo deployment (which we currently are lacking).

File: terracotta/client/app/src/colormap/SinglebandSelector.tsx [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                        activeSinglebandRange !== undefined && 
+                        activeDataset !== undefined &&
+                        (
+                        <>
+                            <Slider 
+                                noNumbers 
+                                getValueCommitted={value => Array.isArray(value) && onSetRangeValue(value)} 
+                                getValue={(value: number | number[]) => Array.isArray(value) && setLocalRange(value)}
+                                defaultValue={activeSinglebandRange} 
+                                min={datasets[activeDataset - page * limit]?.range[0] || 0}
```
Comment (rihr) [inline terracotta/client/app/src/colormap/SinglebandSelector.tsx:88]: Could put in consts

File: terracotta/client/app/src/sidebar-datasets/DatasetPreview.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                                <code>
+                                    {'Metadata\n'}
+                                </code>
+                            </Typography>
+                                <code style={{ whiteSpace: 'pre' }}>
+                                    {'{\n'}
+                                    {Object.keys(dataset).reduce(
+                                        (acc: string[], keyItem: string, j: number) => 
+                                           {
+                                                const neededKeys = ['mean', 'range', 'stdev', 'valid_percentage'];
```
Comment (rihr) [inline terracotta/client/app/src/sidebar-datasets/DatasetPreview.tsx:86]: Could be a function var

File: terracotta/client/app/src/sidebar-datasets/SidebarDatasets.tsx [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+
+    
+    }
+
+    // useEffect(() => {
+
+    //     void getKeys(host)
+
+    // }, [host]) // eslint-disable-line react-hooks/exhaustive-deps
+
```
Comment (rihr) [inline terracotta/client/app/src/sidebar-datasets/SidebarDatasets.tsx:152]:
```suggestion

```

**Review by rihr — APPROVED** [review]

**Comment by dionhaefner** [comment]
Thanks for the implementation, it sure looks more professional now.

I just tried it, and I do have some comments:

- The info text when you click on details seems superfluous now. Why not just remove it? It's hard to find anyway.
- The bold `/apidoc` in the details section looks weird to me.
- Raster listing starts at page 0, and no indication how many pages there are.
- No padding between metadata and preview image looks cramped.
- <s>What happened to the zoom to the dataset? Imaging you have a raster of one yard in New Zealand, how are you ever going to find it on the world map?</s> Just saw that it *does* zoom in, but not if you select the first dataset in the list :thinking:
- The function of the "change base map" button is not obvious before you click on it. Couldn't this be "satellite view" / "street view" with a matching icon instead?
- Sidebar looks very large on smaller / non-maximized windows, and is not resizable anymore:

Screen Shot 2021-04-26 at 10 11 13

- I just saw that you can minimize the sidebar. That arrow is hard to find. Couldn't this be on the left side of the sidebar?
- **RGB view is broken, there is no way to select different rasters for the bands.**
- I don't think the RGB contrast sliders work as they should. They just have a range of 0-255 instead of the range of the rasters.
- Key descriptions are not part of the UI anymore?
- Some more vertical breathing room would be nice.

No one has asked for my review, so you can do with this feedback whatever you want.

**Comment by j08lue** [comment]
Thanks for the ex post facto review, @dionhaefner. 😄 Thought it was nice to push the remake out there, great you are already picking up the revision on another PR, @bertearazvan.

**Pull request merged**